contract/models/abstract_contract_line.py

```diff
@@ -206,18 +208,20 @@ def _compute_price_unit(self):
                         line.contract_id.company_id
                     ).property_product_pricelist
                 )
+                quantity = line.env.context.get("contract_line_qty", line.quantity)
                 product = line.product_id.with_context(
-                    quantity=line.env.context.get(
-                        "contract_line_qty",
-                        line.quantity,
-                    ),
+                    quantity=quantity,
                     pricelist=pricelist.id,
                     partner=line.contract_id.partner_id.id,
                     date=line.env.context.get(
                         "old_date", fields.Date.context_today(line)
                     ),
                 )
-                line.price_unit = pricelist._get_product_price(product, quantity=1)
+                base_price = pricelist._get_product_price(product, quantity=quantity)
+                price = product.uom_id._compute_price(base_price, line.uom_id)
+                line.price_unit = float_round(
+                    price, precision_digits=line.currency_id.decimal_places or 2
```

Comment:
The price unit precision is not the currency one. In fact, you don't have to round anything, as the ORM does it through the `digits` parameter when storing.